### PR TITLE
Add state (i.e. queued, running) tracking for a job lifecycle. Currently...

### PR DESCRIPTION
**Pull request by mngo87**
..., everything is stored in memory. Look into persisting this state onto Cassandra.

### CODE REVIEW

**Comment by mngo87** [comment]
screen shot 2015-03-11 at 3 46 08 pm

This is a screenshot of the UI changes.

**Comment by brndnmtthws** [comment]
This is interesting. All of `JobStats.scala` can probably be refactored to use the method you've added.

File: src/main/scala/org/apache/mesos/chronos/scheduler/mesos/MesosJobFramework.scala [hunk 1]
```diff
@@ -215,6 +215,9 @@ import scala.concurrent.ExecutionContext.Implicits.global
         scheduler.handleFailedTask(taskStatus)
       case TaskState.TASK_RUNNING =>
         log.info("Task with id '%s' RUNNING.".format(taskStatus.getTaskId.getValue))
+      case TaskState.TASK_KILLED =>
```
Comment (brndnmtthws) [inline src/main/scala/org/apache/mesos/chronos/scheduler/mesos/MesosJobFramework.scala]: Nice.

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala [hunk 2]
```diff
@@ -236,4 +306,102 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
...
+        attempt=Some(attempt),
+        isFailure=Some(true))
+  }
+
+  /**
+   * Helper method that performs an insert statement to update the
+   * job statistics (chronos) table. All arguments are surrounded
+   * by options so that a subset of values can be inserted.
+   */
+  private def insertToStatTable(id: Option[String],
```
Comment (brndnmtthws) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: We should really just reuse this for all the methods above, as well.

Comment (mngo87) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: sounds good. will do that and update the pull request.

**Comment by mngo87** [comment]
@brndnmtthws the JobStats has been refactored and tested.  Let me know what you think.

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala [hunk 3]
```diff
@@ -6,70 +6,112 @@ import java.util.logging.{Level, Logger}
...
+  def jobQueued(job: BaseJob, attempt: Int) {
+    updateJobState(job.name, CurrentState.queued)
+  }
+
+  def jobStarted(job: BaseJob, taskStatus: TaskStatus, attempt: Int) {
+    updateJobState(job.name, CurrentState.running)
+    job match {
+      case job: ScheduleBasedJob =>
+        insertToStatTable(
+            id=Some(taskStatus.getTaskId.getValue),
```
Comment (brndnmtthws) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: Could you apply a little more DRY here? A lot of these fields are common between the 2 job types. You only need to get the non-common fields first, then make one call to `insertToStatTable`.

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala [hunk 4]
```diff
@@ -231,9 +324,10 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
         resetSession()
         log.log(Level.WARNING, "No hosts were available, will retry next time.", e)
       case e: QueryExecutionException =>
-        log.log(Level.WARNING, "Query execution failed:", e)
+        log.log(Level.WARNING,"Query execution failed:", e)
```
Comment (brndnmtthws) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: Space after comma.

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala [hunk 5]
```diff
@@ -231,9 +324,10 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
...
       case e: QueryValidationException =>
-        log.log(Level.WARNING, "Query validation failed:", e)
+        log.log(Level.WARNING,"Query validation failed:", e)
```
Comment (brndnmtthws) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: Ditto.

**Comment by brndnmtthws** [comment]
Added a couple more notes.

@elingg any comments?

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala (hunk 3 above)
Comment (elingg) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: +1 on this, combine whatever logic you can outside of the case statement so you are not duplicating code

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala [hunk 6]
```diff
@@ -124,106 +166,157 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
...
-        resetSession()
-        log.log(Level.WARNING, "No hosts were available, will retry next time.", e)
-      case e: QueryExecutionException =>
-        log.log(Level.WARNING, "Query execution failed:", e)
-      case e: QueryValidationException =>
-        log.log(Level.WARNING, "Query validation failed:", e)
+    updateJobState(job.name, CurrentState.idle)
+    job match {
+      case job: ScheduleBasedJob =>
+        insertToStatTable(
```
Comment (elingg) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: A lot of this code is duplicated between the cases. Let's separate out the common logic.

File: src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala [hunk 7]
```diff
@@ -124,106 +166,157 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
...
+            slaveId=Some(taskStatus.getSlaveId.getValue),
+            attempt=Some(attempt),
+            jobSchedule=None, message=None,
+            isFailure=None)
     }
   }
 
   def jobFailed(job: BaseJob, taskStatus: TaskStatus, attempt: Int) {
-    try {
+    updateJobState(job.name, CurrentState.idle)
```
Comment (elingg) [inline src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala]: A lot of this code is duplicated between the cases. Let's separate out the common logic.

**Comment by elingg** [comment]
Nice work! I really like this UI update. I left a few comments.

**Comment by mngo87** [comment]
Thanks @elingg.

@brndnmtthws @elingg, i updated the code to address your comments.  Please review again.

**Comment by brndnmtthws** [comment]
This looks pretty good to me. 2 final notes:
 - please squash this into 1 commit, and
 - how has it been tested?

**Comment by mngo87** [comment]
@brndnmtthws - will squash and reference the new pull request. Cause this is a UI component, this has been tested by myself (and members of my team) by using it.  This feature has been available in my team for a couple months.  I went over scheduled and dependent jobs, as well as test different job durations.  Also tested adding and removing jobs.  Is this sufficient?

**Comment by brndnmtthws** [comment]
You can just force push to this branch with the new commit, rather than opening a new PR: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

WRT to testing, it sounds good to me. I'll let @elingg make the final call on merging.

**Comment by mngo87** [comment]
@brndnmtthws - squashed

**Comment by elingg** [comment]
Much cleaner, the testing sounds good to me. I'm assuming here that it has been tested after the refactoring was done as well.  :+1:

**Comment by mngo87** [comment]
@elingg  - Yes this has been tested after the refactor.

**Pull request merged**